fili-core/src/main/java/com/yahoo/bard/webservice/data/config/table/PhysicalTableDefinition.java

```diff
@@ -1,93 +1,144 @@
-    public PhysicalTableDefinition(
+    protected PhysicalTableDefinition(
             TableName name,
-            ZonedTimeGrain grain,
-            Iterable<? extends DimensionConfig> dimensions
+            ZonedTimeGrain timeGrain,
+            Set<FieldName> metricNames,
+            Set<? extends DimensionConfig> dimensionConfigs
     ) {
         this.name = name;
-        this.grain = grain;
-        this.dimensions = ImmutableSet.copyOf(dimensions);
-        this.logicalToPhysicalNames = Collections.unmodifiableMap(
-                this.dimensions.stream()
-                        .collect(
-                                Collectors.toMap(
-                                        DimensionConfig::getApiName,
-                                        dim -> {
-                                            String physicalName = dim.getPhysicalName();
-                                            if (physicalName == null) {
-                                                throw new RuntimeException("No physical name found for dimension: "
-                                                        + dim.getApiName());
-                                            }
-                                            return physicalName;
-                                        }
-                                )
-                        )
-        );
+        this.timeGrain = timeGrain;
+        this.metricNames = ImmutableSet.copyOf(metricNames);
+        this.dimensionConfigs = ImmutableSet.copyOf(dimensionConfigs);
+        this.logicalToPhysicalNames = Collections.unmodifiableMap(buildLogicalToPhysicalNames(dimensionConfigs));
     }
 
     /**
-     * Define a physical table using a zoneless time grain, defaulting it to UTC.
+     * Get the set of physical tables required to build the current physical table.
+     *
+     * @return a set of physical table names
+     */
+    public abstract Set<TableName> getDependentTableNames();
+
+    /**
+     * Given a set of metric names and a data source metadata service, build the corresponding physical table.
      *
-     * @param name  The table name
-     * @param grain  The zoneless time grain
-     * @param dimensions  The dimension configurations
+     * @param dictionaries  Dictionary containing dimension dictionary and physical table dictionary
+     * @param metadataService  Service containing column available interval information
      *
-     * @deprecated The time zone of a physical table should be set explicitly rather than rely on defaulting to UTC
+     * @return optional of the corresponding type of physical table, empty optional if could not be build.
      */
-    @Deprecated
-    public PhysicalTableDefinition(
-            TableName name,
-            ZonelessTimeGrain grain,
-            Iterable<? extends DimensionConfig> dimensions
-    ) {
-        this(name, grain.buildZonedTimeGrain(DateTimeZone.UTC), dimensions);
-    }
+    public abstract Optional<PhysicalTable> build(
```

Comment:
Not much now, but it feels right that it should return an `Optional` when you build something. Maybe some fili application will extend some new dimension and dimension definition that will fail to build for some reason. Let me know if you want me to remove the `Optional` or not.